File: LeetCode/2353/python/lc2353.py
from collections import defaultdict
from typing import List
from sortedcontainers import SortedSet

# Runtime complexity: O(n).
#  - __init__(): O(n)
#  - changeRating: O(log(n)) on average; O(n) in worst case.
#  - highestRated: O(log(n)).
# Auxiliary space complexity: O(n).
# Subjective level: medium.
class FoodRatings:
    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.foodToRating = dict()
        self.foodToCuisine = dict()
        self.cuisineToCombo = defaultdict(
            lambda: SortedSet(key=lambda k: (-k[0], k[1]))
        )
        for i in range(len(foods)):
            self.foodToRating[foods[i]] = ratings[i]
            self.foodToCuisine[foods[i]] = cuisines[i]
            self.cuisineToCombo[cuisines[i]].add((ratings[i], foods[i]))

    def changeRating(self, food: str, newRating: int) -> None:
        rating = self.foodToRating[food]
        cuisine = self.foodToCuisine[food]
        combo = self.cuisineToCombo[cuisine]
        combo.remove((rating, food))
        combo.add((newRating, food))
        self.foodToRating[food] = newRating

    def highestRated(self, cuisine: str) -> str:
        combo = self.cuisineToCombo[cuisine]
        return combo[0][1]

File: LeetCode/2353/cpp/lc2353.cpp
#include <cassert>
#include <set> // orders elements automatically
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

class FoodRatings {
public:
  FoodRatings(std::vector<std::string>& foods,
              std::vector<std::string>& cuisines,
              std::vector<int>&         ratings)
  {
    for (int i = 0; i < foods.size(); ++i) {
      m_foodToRating[foods[i]] = ratings[i];
      m_foodToCuisine[foods[i]] = cuisines[i];
      m_cuisineToData[cuisines[i]].insert({-ratings[i], foods[i]});
    }
  }

  void changeRating(std::string food, int newRating) {
    const int oldRating = m_foodToRating[food];
    const std::string cuisine = m_foodToCuisine[food];
    auto& fullCuisineData = m_cuisineToData[cuisine];

    fullCuisineData.erase({-oldRating, food});
    fullCuisineData.insert({-newRating, food});
    m_foodToRating[food] = newRating;
  }

  std::string highestRated(std::string cuisine) {
    //return m_cuisineToData[cuisine].rbegin()->second; //rbegin() points to the last element
    // Note: rbegin() approach would get ties wrong. Switched to using negative ratings and begin().
    return m_cuisineToData[cuisine].begin()->second;
  }

private:
  std::unordered_map<std::string, int> m_foodToRating;
  std::unordered_map<std::string, std::string> m_foodToCuisine;
  // note: has to be <rating, name> instead of <name, rating> so that automatic sorting works as required.
  std::unordered_map<std::string, std::set<std::pair<int, std::string>>> m_cuisineToData;
};

int main() {
  std::vector<std::string>    foods = {"kimchi",     "miso",    "sushi",  "moussaka",    "ramen", "bulgogi"};
  std::vector<std::string> cuisines = {"korean", "japanese", "japanese",     "greek", "japanese",  "korean"};
  std::vector<int>          ratings = {       9,         12,          8,          15,         14,         7};

  FoodRatings fr(foods, cuisines, ratings);
  assert(fr.highestRated("korean") == "kimchi");
  assert(fr.highestRated("japanese") == "ramen");
  fr.changeRating("sushi", 16);
  assert(fr.highestRated("japanese") == "sushi");
  fr.changeRating("ramen", 16);
  assert(fr.highestRated("japanese") == "ramen"); // lexicographically smaller than "sushi".
}
